rust/wasmtime_bindings/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+        fn create_store(engine: &mut Engine) -> Result<Box<Store>>;
+
+        type Memory;
+        fn data(self: &Memory, store: &Store) -> *mut u8;
+        fn size(self: &Memory, store: &Store) -> u64;
+        fn grow(self: &mut Memory, store: &mut Store, delta: u64) -> Result<u64>;
+        fn get_memory(instance: &Instance, store: &mut Store) -> Result<Box<Memory>>;
+        fn get_abi(instance: &Instance, store: &mut Store, memory: &Memory) -> Result<u32>;
+
+        type Engine;
+        fn create_engine() -> Result<Box<Engine>>;
+
+        type Func;
+        fn create_func(
+            instance: &Instance,
+            store: &mut Store,
+            function_name: &str,
+        ) -> Result<Box<Func>>;
+
+        type Val;
+        fn kind(self: &Val) -> Result<ValKind>;
+        fn i32(self: &Val) -> Result<i32>;
+        fn i64(self: &Val) -> Result<i64>;
+        fn f32(self: &Val) -> Result<f32>;
+        fn f64(self: &Val) -> Result<f64>;
+
+        // vector of wasmtime values, used for passing arguments to a function and returning results from a function
+        type ValVec;
+        fn push_i32(self: &mut ValVec, val: i32) -> Result<()>;
+        fn push_i64(self: &mut ValVec, val: i64) -> Result<()>;
+        fn push_f32(self: &mut ValVec, val: f32) -> Result<()>;
+        fn push_f64(self: &mut ValVec, val: f64) -> Result<()>;
+        fn pop_val(self: &mut ValVec) -> Result<Box<Val>>;
+        fn get_val_vec() -> Result<Box<ValVec>>;
+
+        // type and methods for returning and executing a WebAssembly function asynchronously
+        type Fut<'a>;
+        unsafe fn resume<'a>(self: &mut Fut<'a>) -> Result<bool>;
+        unsafe fn get_func_future<'a>(
+            store: &'a mut Store,
+            func: &'a Func,
+            args: &'a ValVec,
+            rets: &'a mut ValVec,
+        ) -> Result<Box<Fut<'a>>>;
+    }
+}
+
+pub struct Instance {
+    wasmtime_instance: wasmtime::Instance,
+}
+
+fn create_instance(engine: &Engine, module: &Module, store: &mut Store) -> Result<Box<Instance>> {
+    let mut linker = wasmtime::Linker::new(&engine.wasmtime_engine);
+    if let Err(e) = wasmtime_wasi::tokio::add_to_linker(&mut linker, |s| s) {
+        return Err(anyhow!("Failed to add wasi to linker: {:?}", e));
+    }
+    let mut inst_fut =
+        Box::pin(linker.instantiate_async(&mut store.wasmtime_store, &module.wasmtime_module));
+    let waker = futures::task::noop_waker_ref();
+    let mut ctx = Context::from_waker(&waker);
+
+    loop {
+        match inst_fut.as_mut().poll(&mut ctx) {
+            Poll::<Result<wasmtime::Instance>>::Pending => {}
+            Poll::<Result<wasmtime::Instance>>::Ready(Ok(inst)) => {
+                return Ok(Box::new(Instance {
+                    wasmtime_instance: inst,
+                }));
+            }
+            Poll::<Result<wasmtime::Instance>>::Ready(Err(e)) => {
+                return Err(anyhow!("Failed to instantiate module: {:?}", e));
+            }
```

Comment:
Why does instantiate_async() return a future?

If it's in order to yield, then aren't we wasting the opportunity by looping? We should return to the caller and let them yield too.

Reply:
(instantiating an instance is less important than execution, since it's rarer, but still)

Reply:
According to the documentation, we return a future here only to potentially yield when we "asynchronously invoke the wasm start function in case it calls any imported function which is an asynchronous host function". The only imports come from WASI, but I don't think they are used in the start function (actually I couldn't find the start function implementation - it is also possible that there is none).
If we also go with https://github.com/scylladb/scylladb/pull/11351#discussion_r962925721 no imports will be async at all, so it won't be possible that this future yields in the middle